### PR TITLE
adding the calculation and rendering of gpa weighted by number of lines per file

### PR DESCRIPTION
**Pull request by natesholland**
Ok I rebased this all into one commit so it should be ready to come in. If you see anything you don't like please let me know.

### CODE REVIEW

File: lib/rubycritic/core/analysed_module.rb [hunk 1]
```diff
@@ -14,6 +14,7 @@ class AnalysedModule
     attribute :complexity
     attribute :duplication,   Integer, :default => 0
     attribute :methods_count
+    attribute :lines
```
Comment (guilhermesimoes) [inline lib/rubycritic/core/analysed_module.rb]: Must have missed this before, but I believe this line is not needed.

File: lib/rubycritic/generators/global_rating_calculator.rb [hunk 2]
```diff
@@ -0,0 +1,15 @@
+module Rubycritic
+  module Generator
+    module GlobalRatingCalculator
+      def calculate_gpa(analysed_modules)
+        gpa_sum = 0
+        total_lines = 0
+        analysed_modules.each do |analysed_module|
+          gpa_sum += (analysed_module.rating.to_gpa) * analysed_module.lines
```
Comment (guilhermesimoes) [inline lib/rubycritic/generators/global_rating_calculator.rb]: What's the reasoning for converting the rating back to a GPA step? I think I like #35's approach better, averaging the costs of the modules and then using the Rating class to convert it to a GPA.

File: lib/rubycritic/generators/html_report.rb [hunk 3]
```diff
@@ -3,15 +3,18 @@
...
 module Rubycritic
   module Generator
 
     class HtmlReport
+      include GlobalRatingCalculator
       ASSETS_DIR = File.expand_path("../html/assets", __FILE__)
 
       def initialize(analysed_modules)
         @analysed_modules = analysed_modules
+        @gpa = calculate_gpa(analysed_modules)
```
Comment (guilhermesimoes) [inline lib/rubycritic/generators/html_report.rb]: I don't think it's a `Report`'s responsibility to know how to calculate the GPA. Make `GlobalRatingCalculator` a class and call this method on it!

File: lib/rubycritic/generators/global_rating_calculator.rb (hunk 2 above)
Comment (natesholland) [inline lib/rubycritic/generators/global_rating_calculator.rb]: I think converting the rating back to a GPA makes sense because it will put a cap on how much a really bad file can affect the total GPA. Since the rating can have an arbitrarily high cost it means that when averaged together a really bad F can skew the entire rating. I know at Spiceworks we have a model that is very very bad (over 4000 lines of code) so it would skew our GPA lower than if we made all F's the same.

File: lib/rubycritic/core/analysed_module.rb (hunk 1 above)
Comment (natesholland) [inline lib/rubycritic/core/analysed_module.rb]: This line is needed if we want to have the lines= method which is very useful in testing. I can pull it out if you want and then rework the tests, but I think it's worth leaving in for testing purposes.

File: lib/rubycritic/generators/html_report.rb (hunk 3 above)
Comment (natesholland) [inline lib/rubycritic/generators/html_report.rb]: Agreed! And done.

**Comment by natesholland** [comment]
It turned out that making the weighting by lines of code seemed to have little effect on the total time needed to calculate the GPA, so I went ahead and weighted by lines of code instead of lines in the file.

**Comment by natesholland** [comment]
@guilhermesimoes can you pick this or #35 and merge one of them in? I think there are enough people who want these changes to warrant merging them in.

**Comment by natesholland** [comment]
@guilhermesimoes any more thoughts on this? I really don't want to fork this internally at Spiceworks but I am probably going to if we can't pick some form of internal global rating. I'd be happy to do extra work on this to get it merged in if there is anything you are waiting on. Please let me know if you intent to merge this in.

**Comment by JustinAiken** [comment]
:+1: for picking one - no opinion as to which one.

**Comment by natesholland** [comment]
@Onumis and @troessner can we get this merged into the main trunk of rubycritic?

**Comment by troessner** [comment]
@natesholland i have "review all open pull requests" on my to do list already. Please give me / us some time to become fully acquainted with the project.

**Comment by natesholland** [comment]
Sounds good, I just wanted to make sure it was in the queue

File: lib/rubycritic/core/analysed_module.rb [hunk 4]
```diff
@@ -27,6 +28,15 @@ def rating
       @rating ||= Rating.from_cost(cost)
     end
 
+    def lines
```
Comment (Onumis) [inline lib/rubycritic/core/analysed_module.rb]: `lines_of_code` would be a more appropriate name for this.

**Comment by Onumis** [comment]
Ok so, I'm gonna have to lay some though on this and #35, there are a lot of really interesting opinions here and there and I'm not convinced on which one is the "right path".

File: lib/rubycritic/core/analysed_module.rb [hunk 5]
```diff
@@ -27,6 +28,15 @@ def rating
...
+      return @lines if @lines
```
Comment (troessner) [inline lib/rubycritic/core/analysed_module.rb]: You could use the more idiomatic memoization pattern:

```Ruby
@lines ||= begin
  .....
end
```

File: lib/rubycritic/core/analysed_module.rb [hunk 6]
```diff
@@ -27,6 +28,15 @@ def rating
...
+      count = 0
+      File.open(path).each_line do |line|
+        count += 1 unless line.strip[0] == "#" || line.strip.empty?
```
Comment (troessner) [inline lib/rubycritic/core/analysed_module.rb]: Let's move the guard

```Ruby
line.strip[0] == "#" || line.strip.empty?
```

into an appropriately named method.
How about

```Ruby
def ignore_line?(line)
  line.strip[0] == "#" || line.strip.empty?
end
```

?

You could directly use File's `readlines` and Enumerable's `reject`, then this whole method could be written more concise like this:

```Ruby
def lines
  @lines ||= File.open(path).readlines
                                 .reject { |line| ignore_line?(line) }
                                  .size
```

Then you can also drop the "begin" I suggested above ;)

File: lib/rubycritic/core/rating.rb [hunk 7]
```diff
@@ -25,6 +25,17 @@ def to_h
     def to_json(*a)
       to_h.to_json(*a)
     end
+
+    def to_gpa
+      if @letter == "A" then 4.0
+      elsif @letter == "B" then 3.0
+      elsif @letter == "C" then 2.0
+      elsif @letter == "D" then 1.0
+      elsif @letter == "F" then 0.0
```
Comment (troessner) [inline lib/rubycritic/core/rating.rb]: Let's not mix up code with data.
How about putting the association of character to number into a hash?

```Ruby
{ 'A' => 4.0, ....}
````
Then you could just use Hash#fetch on that.

File: lib/rubycritic/generators/global_rating_calculator.rb [hunk 8]
```diff
@@ -0,0 +1,15 @@
+module Rubycritic
+  module Generator
+    class GlobalRatingCalculator
```
Comment (troessner) [inline lib/rubycritic/generators/global_rating_calculator.rb]: You never plan to instantiate this, right?
Then let's rather use a module and `module_function` instead of self :)

File: test/lib/rubycritic/report_generators/global_rating_calculator_test.rb [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+require "rubycritic/core/analysed_module"
+
+describe Rubycritic::Generator::GlobalRatingCalculator do
+  it "calculates the gpa" do
+    smells1 = [SmellDouble.new(:cost => 1), SmellDouble.new(:cost => 2)]
+    smells2 = [SmellDouble.new(:cost => 10), SmellDouble.new(:cost => 12)]
+    am1 = Rubycritic::AnalysedModule.new(:smells => smells1, :complexity => 0, :lines => 25)
+    am2 = Rubycritic::AnalysedModule.new(:smells => smells2, :complexity => 0, :lines => 10)
+    analysed_modules = [am1, am2]
+    gpa = Rubycritic::Generator::GlobalRatingCalculator.calculate_gpa(analysed_modules)
```
Comment (troessner) [inline test/lib/rubycritic/report_generators/global_rating_calculator_test.rb]: The test set up here is huge. Normally I'd suggest to introduce factories here but I guess this would be a bigger topic that should be adressed in an separate pull request.

File: lib/rubycritic/generators/global_rating_calculator.rb [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+      def self.calculate_gpa(analysed_modules)
+        gpa_sum = 0
+        total_lines = 0
+        analysed_modules.each do |analysed_module|
+          gpa_sum += (analysed_module.rating.to_gpa) * analysed_module.lines
```
Comment (troessner) [inline lib/rubycritic/generators/global_rating_calculator.rb]: The braces dont make sense here, seems like a leftover?

**Comment by troessner** [comment]
Closing this one since we merged #35 - thanks!
